### PR TITLE
Use span of impl/trait in len_without_is_empty error message, rather …

### PR DESCRIPTION
**Pull request by theotherphil**
…than the span of the len method. (Fixes https://github.com/Manishearth/rust-clippy/issues/1532)

### CODE REVIEW

File: tests/ui/len_zero.stderr [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
-10 |       pub fn len(self: &Self) -> isize {
-   |  _____^ starting here...
+9  |   impl PubOne {
+   |  _^ starting here...
+10 | |     pub fn len(self: &Self) -> isize {
 11 | |         1
 12 | |     }
-   | |_____^ ...ending here
+13 | | }
+   | |_^ ...ending here
```
Comment (mcarton) [inline tests/ui/len_zero.stderr]: This makes me wonder what happens when there is more than one `impl` for a type.

Can you also had a test that checks that this actually fixes #1532?

Comment (theotherphil) [inline tests/ui/len_zero.stderr]: Good point. The current lint has a false positive if `len` and `is_empty` are defined in separate impl blocks.

Comment (theotherphil) [inline tests/ui/len_zero.stderr]: I've pushed a change which:

* adds an extra, irrelevant impl block to PubOne to check that we this doesn't generate any errors
* adds a PubAllowed struct equivalent to PubOne but with an `allow` attribution on the relevant impl.

Are you happy to merge this as is, or do you want to wait for a fix to the impls-in-different-blocks issue?

**Comment by oli-obk** [comment]
thanks!

File: tests/ui/len_zero.stderr (hunk 1 above)
Comment (oli-obk) [inline tests/ui/len_zero.stderr]: > Are you happy to merge this as is, or do you want to wait for a fix to the impls-in-different-blocks issue?

That's a different issue. I reported it as #1562

**Pull request merged**